trufflesuite/truffle
# Implement console.log in Truffle

## PR description

Integrates [`@ganache/console.sol`](https://github.com/trufflesuite/ganache/tree/develop/src/chains/ethereum/console.log) Smart contract logging into Truffle's test and develop workflow. 

Note: Ganache already provides console logging, but it isn't integrated into Truffle (if you start a ganache instance and use truffle to interact with your smart contract on the ganache instance, logging output will be in the Ganache's TTY)

This is part of the effort outlined in #5282 and should address: #5193 and #5195

<details><summary>Click to see sample output using truffle test</summary>

Note every line prefixed with a `:` is a console.log output.

```console
$ truffle test
Using network 'test'.


Compiling your contracts...
===========================
> Everything is up to date, there is nothing to compile.
  : CSTR::Simple
  :     secret: 13n


  Contract: Simple
    Works with migrations deployment
      ✔ Should have a deployed contract
      ✔ has the secret
  : The passphrase is: 13n
  : Leak it a hundred times, plus one more: 1313n
      ✔ has the derived Secret


    Works with new deployment
  : CSTR::Simple
  :     secret: 551n
      ✔ Should have a deployed contract
      ✔ has the secret
  : The passphrase is: 551n
  : Leak it a hundred times, plus one more: 55651n
      ✔ has the derived Secret


  6 passing (139ms)

```

</details>


### How to use in a project

- Import `truffle/Console.sol` in your contract and use its interface. See [Ganache docs for usage details](https://github.com/trufflesuite/ganache/tree/develop/src/chains/ethereum/console.log)
  ```solidity
   pragma solidity >=0.4.25 <0.9.0;
   import "truffle/Console.sol";
   
   ...
   // in some function...
   console.log("hello msg.sender: %o", msg.sender);
   ```
- Configuration: There are 2 options for configuration, both set in `truffle-config.js`. 
  1. `solidityLog.displayPrefix` sets the prefix for every console.log line. The
     default is `""`; however, it is a nice option when you want to differentiate
     console.log output from extra verbose text output.
  2. `solidityLog.preventConsoleLogMigration` a safety measure to disable accidentally
     migrating contracts that use console.log to MAINNET. The default is `false`,
     you would have to opt in, setting to `true`, should you need it. 

## Testing instructions

This is more of a smoke test scenario. Try adding console.log in a contract, or if you prefer, [try one of these](https://github.com/cds-amal/console-log-tests). Does logging work? Is it intuitive? Do reverts log how you expect? Please try to break it. This feature only works when using `truffle develop` or `truffle test`. 


- To test `preventConsoleLogMigration` you will need to run the project against a network that simulates Mainnet. This can be done by starting ganache w/ `ganache  --chain.networkId 1`

  The linked projects define a `mainfork` network in their `truffle-config.js`,
  so specify it in the migrate command (`truffle migrate --network mainfork`).

  |mainnet | preventConsoleLogMigration| expectation|
  |---|---|---|
  |false |false| it should deploy|
  |false |true| it should deploy |
  |true |false| it should deploy|
  |true |true| it should NOT deploy, and print a helpful, actionable message.|
  
  1. Start ganache w/ `ganache --chain.networkId 1` (simulate mainnet)
  2. Set `preventConsoleLogMigration` appropriately. test true/false on mainnet and not mainnet
  2. `truffle migrate --network mainfork` for testing on mainnet, if using one the linked projects.


## Documentation


## Breaking changes and new features
